[Security Solution] Elastic Security Assistant

spong opened the pull request:
## [Security Solution] Elastic Security Assistant

The _Elastic Security Assistant_ has entered the chat, integrating generative AI and large language models (LLMs) into the workflows of Elastic Security users.

Bring your alerts, events, rules, and data quality checks into the conversation.

<https://github.com/elastic/kibana/assets/2946766/31d65c78-5692-4817-b726-820c5df0801c>

This PR merges a feature branch developed by @spong and @andrew-goldstein , seeded by @jamesspi 's prototype of the assistant. Connectivity to LLMs is provided the [Generative AI Connector](<https://github.com/elastic/kibana/pull/157228>) , developed by @stephmilovic . This PR includes:

- A new reusable Kibana package containing the assistant: `x-pack/packages/kbn-elastic-assistant`
  - See the `How to embed the Assistant in other parts of Kibana` for details
- Assistant integration into Elastic Security Solution workflows (e.g. alerts, cases, Timeline, rules, data quality)

### An assistant trained on the Elastic stack and Elastic Security

The [Generative AI Connector](<https://github.com/elastic/kibana/pull/157228>) connects the assistant to OpenAI and Azure OpenAI models trained with knowledge of the Elastic stack and the Elastic Security solution, including:

- The Elastic open [Detection Rules](https://github.com/elastic/detection-rules)
- The [Elastic Common Schema (ECS)](https://www.elastic.co/guide/en/ecs/current/index.html)
- Elastic query languages, including [KQL](https://www.elastic.co/guide/en/kibana/current/kuery-query.html), [EQL](https://www.elastic.co/guide/en/elasticsearch/reference/current/eql-syntax.html), and the [Elastic Query DSL](https://www.elastic.co/guide/en/elasticsearch/reference/current/query-dsl.html)
- [Elasticsearch API documentation](https://www.elastic.co/guide/en/elasticsearch/reference/8.8/rest-apis.html)

This training enables the assistant to offer fully interactive chat experiences that include:

- alert summarization
- interactive query generation
- workflow suggestions
- generating ingestion configurations that conform to the Elastic Common Schema
- your imagination

using context from Elastic Security.

### Take action from your conversations

The Actions (from assistant response):

- Send KQL to Timeline
- Send EQL to Timeline
- Send Elasticsearch DSL to Timeline
- Send Note to timeline
- Create new case
- Add to existing case
- Copy to clipboard

### Feature Flag

This functionality has been introduced behind the `assistantEnabled` feature flag. Please add this configuration to your `kibana.yml` or Kibana Cloud User Settings configuration to enable:

```
xpack.securitySolution.enableExperimental: ['assistantEnabled']
```

### Components architecture diagram

![Untitled Diagram drawio (1)](https://github.com/elastic/kibana/assets/55110838/8f446313-629f-4646-b44d-0cb0ca74aeaf)

### How to embed the Assistant in other parts of Kibana

Follow the general instructions in `x-pack/packages/kbn-elastic-assistant/index.ts` to integrate the assistant into a Kibana app.

#### Step 1 - Wrap your Kibana app in the `AssistantProvider` component

```ts
// Step 1: Wrap your Kibana app in the `AssistantProvider` component. This typically
// happens in the root of your app. Optionally provide a custom title for the assistant:

/** provides context (from the app) to the assistant, and injects Kibana services, like `http` */
export { AssistantProvider } from './impl/assistant_context';
```

#### Step 2: Add the `AssistantOverlay` component to your app

```ts
// Step 2: Add the `AssistantOverlay` component to your app. This component displays the assistant
// overlay in a modal, bound to a shortcut key:

/** modal overlay for Elastic Assistant conversations */
export { AssistantOverlay } from './impl/assistant/assistant_overlay';

// In addition to the `AssistantOverlay`, or as an alternative, you may use the `Assistant` component
// to display the assistant without the modal overlay:

/** this component renders the Assistant without the modal overlay to, for example, render it in a Timeline tab */
export { Assistant } from './impl/assistant';
```

#### Step 3: Wherever you want to bring context into the assistant, use the any combination of the following

```ts
// Step 3: Wherever you want to bring context into the assistant, use the any combination of the following
// components and hooks:
// - `NewChat` component
// - `NewChatById` component
// - `useAssistantOverlay` hook

/**
 * `NewChat` displays a _New chat_ icon button, providing all the context
 * necessary to start a new chat. You may optionally style the button icon,
 * or override the default _New chat_ text with custom content, like `🪄✨`
 *
 * USE THIS WHEN: All the data necessary to start a new chat is available
 * in the same part of the React tree as the _New chat_ button.
 */
export { NewChat } from './impl/new_chat';

/**
 * `NewChatByID` displays a _New chat_ icon button by providing only the `promptContextId`
 * of a context that was (already) registered by the `useAssistantOverlay` hook. You may
 * optionally style the button icon, or override the default _New chat_ text with custom
 * content, like {'🪄✨'}
 *
 * USE THIS WHEN: all the data necessary to start a new chat is NOT available
 * in the same part of the React tree as the _New chat_ button. When paired
 * with the `useAssistantOverlay` hook, this option enables context to be be
 * registered where the data is available, and then the _New chat_ button can be displayed
 * in another part of the tree.
 */
export { NewChatById } from './impl/new_chat_by_id';

/**
 * `useAssistantOverlay` is a hook that registers context with the assistant overlay, and
 * returns an optional `showAssistantOverlay` function to display the assistant overlay.
 * As an alterative to using the `showAssistantOverlay` returned from this hook, you may
 * use the `NewChatById` component and pass it the `promptContextId` returned by this hook.
 *
 * USE THIS WHEN: You want to register context in one part of the tree, and then show
 * a _New chat_ button in another part of the tree without passing around the data, or when
 * you want to build a custom `New chat` button with features not not provided by the
 * `NewChat` component.
 */
export { useAssistantOverlay } from './impl/assistant/use_assistant_overlay';
```

Co-authored-by: Garrett Spong <garrett.spong@elastic.co>
Co-authored-by: Andrew Macri <andrew.macri@elastic.co>


<!--ONMERGE {"backportTargets":["8.8"]} ONMERGE-->

spong committed:
Hiiiiii world

andrew-goldstein committed:
[Security Solution] Security Assistant: useSecurityAssistantQuery hook and New chat button #5

- adds the useSecurityAssistantQuery hook
- adds a `New chat` button that renders a query in a popover

spong committed:
Adds overlay assistant, new quick prompt pills, and cleans up assistant logic

spong committed:
Merge branch 'main' of github.com:elastic/kibana into security-assistant-playground

spong committed:
Adds conversation selector, custom localstorage keys, and UI cleanup

andrew-goldstein committed:
[Security Solution] Security Assistant: context provider hooks #6

- adds hooks that provide context to the Security Assistant

spong committed:
Removes self from state jail

spong committed:
Extracts and fixes conversation selector and adds createConversation helper

spong committed:
Brings back localstorage support and updates localstorage hook to support prev

spong committed:
Adds support for autofocusing prompt when opening timeline or selecting the assistant tab

spong committed:
Adds new useSecurityAssistantOverlay hook for programmatically opening overlay with conversationId and promptContextId

andrew-goldstein committed:
[Security Solution] Security Assistant: Prompt Editor / Sends System Prompts / Sends Prompt Context #7

- adds the query editor
- passes context and system prompt to the api

andrew-goldstein committed:
- previews the prompt context before sending it (#8)

spong committed:
Adds new markdown parser for send to timeline actions for eql, kql and queryDSL

spong committed:
Factored out codeblock analysis, fixed styling in timeline, extracted quick prompts and wired them back up to the prompt textarea

andrew-goldstein committed:
[Security Solution] Security Assistant: Add note to Timeline #9

- Add note to Timeline
- Common New Chat button for alert flyout and Timeline
- Updated context descriptions

spong committed:
Adds selected rules as prompt context, fixes some styling and ux

spong committed:
Fixes data issue with translations

spong committed:
Fixes sent to timeline action render issue

andrew-goldstein committed:
[Security Solution] Security Assistant: fixes i18n errors (#10)

- fixes i18n errors (expressions are not allowed in template literals)

andrew-goldstein committed:
[Security Solution] Security Assistant: bootstraps the @kbn/elastic-assistant package #11

- bootstraps the `@kbn/elastic-assistant` package in `x-pack/packages/kbn-elastic-assistant`

andrew-goldstein committed:
[Security Solution] Security Assistant: Moves the assistant to the @kbn/elastic-assistant package #12

## [Security Solution] Security Assistant: Moves the assistant to the `@kbn/elastic-assistant` package

- moves most assistant code to `x-pack/packages/kbn-elastic-assistant`

spong committed:
Merge branch 'main' of github.com:elastic/kibana into security-assistant-playground

andrew-goldstein committed:
[Security Solution] Security Assistant: fixes linter issues (#13)

- fixes linter issues remaining from ON week

andrew-goldstein committed:
[Security Solution] Security Assistant: take out the trash (icon) (#14)

- replaces the `trash` icon with the `cross`

spong committed:
Merge branch 'main' of github.com:elastic/kibana into security-assistant-playground

spong committed:
Integrates with GenAI Connector

spong committed:
Merge branch 'main' of github.com:elastic/kibana into security-assistant-playground

andrew-goldstein committed:
[Security Solution] Security Assistant: Reduces the public API surface & improves code coverage (#15)

## [Security Solution] Security Assistant: Reduces the public API surface & improves code coverage

- Reduces the public API surface
- improves code coverage

spong committed:
Merge branch 'main' of github.com:elastic/kibana into security-assistant-playground

spong committed:
Removes export of OpenAiProviderType that was added in main

spong committed:
Swathe of usability and bugfixes around connector integration

spong committed:
Merge branch 'main' of github.com:elastic/kibana into security-assistant-playground

andrew-goldstein committed:
[Security Solution] Security Assistant: Data Quality dashboard _New Chat_ button & API updates (#16)

### Summary

- adds a _New chat_ button to the Data Quality dashboard
- updates the `useAssistantOverlay` API
- `NewChat` takes an optional `children` prop
- added `NewChatById` component
- improves test coverage

spong committed:
Merge branch 'main' of github.com:elastic/kibana into security-assistant-playground

YulNaumenko commented:
There is a hidden flyout issue when I'm trying to create Case based on the Assistant response:

https://github.com/elastic/kibana/assets/55110838/f7f1cf59-a9af-4b10-80f9-ca6cfb3b23b0

YulNaumenko commented:
In Timeline Assistant each time I asked the question or get the response, the conversation always jumps to the top:

https://github.com/elastic/kibana/assets/55110838/286292c7-7231-4270-b4c1-b787b2fdb70f

We should keep the end of the conversation active pointer.

spong committed:
Adds CRUD for QuickPrompts and extracts them and base PromptContexts to solution code

spong committed:
Merge branch 'main' of github.com:elastic/kibana into security-assistant-playground

YulNaumenko commented:
Connector switch doesn't seem to be working for Alert assistant in "timelines" context:
 

https://github.com/elastic/kibana/assets/55110838/73f48584-01c1-4ba1-aded-604af1614a7d

spong committed:
i18n updates from James

YulNaumenko commented:
I think we should disable sending message if any connector is selected or at least show the error message with a more descriptive information:

https://github.com/elastic/kibana/assets/55110838/fe69f552-fd6a-4db0-b6e7-7b15d00f24fa

YulNaumenko commented on x-pack/packages/kbn-elastic-assistant/impl/assistant/api.tsx:
```diff
@@ -0,0 +1,133 @@
+  try {
+    // TODO: Find return type for this API
+    // eslint-disable-next-line @typescript-eslint/no-explicit-any
+    const response = await http.fetch<any>(
```
It is fine for the first iteration, but as an alternative to the connector API we can use `ActionsClient` and I prefer it better, because it will be easier to tune its methods to the custom needs like bringing streams functionality.

I was thinking maybe that will be more extendable architecture if that logic will be outside of that package. And `handleSendMessage` could be a consumer callback with the Connector way integration or some other method for sending messages.

spong replied:
Oh gosh, I totally missed there was a client! Thank you -- this will be much preferred 👍. Made a note to swap to the client (but probably won't do in this initial PR if that's okay).

andrew-goldstein committed:
[Security Solution] Security Assistant: adds the assistantEnabled feature flag (#17)

## Summary

- adds the `assistantEnabled` feature flag (default: `false`) in `x-pack/plugins/security_solution/common/experimental_features.ts`

spong committed:
Fixes from PR review and desk testing

spong committed:
Merge branch 'main' of github.com:elastic/kibana into security-assistant-playground

spong commented:
Thanks for the feedback so far @YulNaumenko! @andrew-goldstein and I have added a ton of fixes in [d527032](https://github.com/elastic/kibana/pull/156933/commits/d52703217e98a0a8bf0d500f8fee87a3debdb464), as for your comments:


> There is a hidden flyout issue when I'm trying to create Case based on the Assistant response:

Yeah, looks to be the case, in investigating it seems we can't get a callback when the user selected 'New case' and so can't 
automatically dismiss the assistant overlay. So we'll either need to add support there, or an imperative solution. 


> In Timeline Assistant each time I asked the question or get the response, the conversation always jumps to the top:

Resolved w/ a temporary workaround. Looks like the container is resulting in a new component, which appears to be circumventing the existing logic to auto-scroll to the bottom. 

> Connector switch doesn't seem to be working for Alert assistant in "timelines" context:

So we weren't able to reproduce exactly (we got some other fixes in before), but have improved the UX around the 'empty/missing connector' states, so this should be much better and informative to the user when a connector is missing.

> I think we should disable sending message if any connector is selected or at least show the error message with a more descriptive information:

Yup! As with the item before, we've added a 'connector missing' callout, and ensure it is displayed whenever a conversation either has no connector configured, or references a deleted connector.

spong committed:
Updating api error language

spong committed:
Merge branch 'main' of github.com:elastic/kibana into security-assistant-playground

MadameSheema commented:
**BUG:** Generative AI connector window displayed twice

**Initial setup:**
- No connectors configured

**Steps to reproduce:**
1. Open the `Elastic security assistant`
2. Press the space-bar to skip 
3. Start typing the connector name

**Current behaviour:**
- Another Generative AI connector window is opened

[Alerts - Kibana.webm](https://github.com/elastic/kibana/assets/17427073/73f2130b-c312-4149-b090-c84991de1e68)

MadameSheema commented:
**FEEDBACK/BUG:** The assistant is available in the preview alert flyout. On that flyout we don't allow the user to take any action on the displayed alert. From the assistant we are allowing the user to take action on it. So we should either don't display the `New chat` icon in that flyout, or don't allow the user to take any action from the assistant when it has been opened from the preview flyout.

andrew-goldstein committed:
[Security Solution] Security Assistant: fixes project lint errors  (#18)

## Summary

- fixes project checks via `node ./scripts/lint_ts_projects.js --fix`

MadameSheema commented:
**BUG:** Assistant displayed for a user without connectors authorization

**Initial setup:**
- To have an instance with at least one AI connector
- To have a user without connectors authorization

**Steps to reproduce:**
1. Navigate to any of the pages where we allow to access the assistant

**Current behaviour:**
- The links/icons to acess the assistant are displayed
- The assistant can be opened
- The window to create a new connector is displayed but cannot be filled

**Accessing the assistant from timeline:**
[Timelines - Kibana.webm](https://github.com/elastic/kibana/assets/17427073/da9138b4-ddb8-42fe-a1f2-3427723b7234)

**Accessing the assistant from alert flyout:**
[Alerts - Kibana (2).webm](https://github.com/elastic/kibana/assets/17427073/61dda00b-70b5-4947-be7b-6f3bd79c2c88)

MadameSheema commented:
**QUESTION:** It is expected for conversations to be shared between different spaces?

andrew-goldstein committed:
[Security Solution] Security Assistant: updates CODEOWNERS (#19)

## Summary

- updated `.github/CODEOWNERS` with `node scripts/generate codeowners`

MadameSheema commented:
**BUG FOUND BY CYPRESS:**
**Failing test:**
- Alerts cell actions Add to timeline should add an empty property to default timeline (alerts_cell_actions_cy.ts)

**Issue:**
- Connector window is opened when a space is pressed on the global filter bar.

**Steps to reproduce:**
1. Enter a value and press enter

**Current behaviour:**
- The generative AI connector is displayed

![image](https://github.com/elastic/kibana/assets/17427073/eed3c387-1b4f-4f1d-8484-cbf31f3e6815)

spong committed:
Removes space key listener to resolve cypress issues

andrew-goldstein committed:
[Security Solution] Security Assistant: updates mocks (#20)

## Summary

- updates mocks to fix Jest tests

YulNaumenko commented on x-pack/plugins/security_solution/public/app/app.tsx:
```diff
@@ -17,18 +18,24 @@ import type { AppLeaveHandler, AppMountParameters } from '@kbn/core/public';
+
+import { getComments } from '../assistant/get_comments';
+import { augmentMessageCodeBlocks, LOCAL_STORAGE_KEY } from '../assistant/helpers';
+import { useConversationStore } from '../assistant/use_conversation_store';
```
I'd prefer to use package export for this. Is there any reason why we are not able to export it with package?

spong replied:
Yeah, so this `useConversationStore` is a 'solution specific' hook for managing conversation state outside of the assistant. So it is separate from the `useConversation` hook part of the assistant package api. We can clarify this in the documentation (and hopefully this is only temporary while we use localstorage).

spong committed:
Test fixes, i18n update, and additional feature flag conditional

spong committed:
Merge branch 'main' of github.com:elastic/kibana into security-assistant-playground

YulNaumenko commented:
I'm not able any more to select connector and make assistant available:

https://github.com/elastic/kibana/assets/55110838/501e9d22-d5ca-4567-bff5-c426226fdf00

YulNaumenko commented:
Creating new connector doesn't help:

https://github.com/elastic/kibana/assets/55110838/b24d4b5d-1a10-46b0-b9e4-8f01ed9a9196

YulNaumenko commented:
I realized I need to type something before send message button will be available, so Promps only don't work:

https://github.com/elastic/kibana/assets/55110838/c0d18bee-cceb-4cc2-bfbc-169d5f145410

spong committed:
Fixes prompt submit issue

spong committed:
Merge branch 'main' of github.com:elastic/kibana into security-assistant-playground

spong committed:
Skips flakey cypress test

YulNaumenko commented on x-pack/packages/security-solution/ecs_data_quality_dashboard/tsconfig.json:
```diff
@@ -22,5 +22,7 @@
     "@kbn/core-http-browser",
     "@kbn/core-http-browser-mocks",
+    "@kbn/elastic-assistant",
+    "@kbn/triggers-actions-ui-plugin",
```
Why do we need the reference for `triggers-actions-ui-plugin` here?

spong replied:
IIRC it was needed because of the mock for tests.

spong committed:
Merge branch 'main' of github.com:elastic/kibana into security-assistant-playground

YulNaumenko commented on x-pack/packages/security-solution/ecs_data_quality_dashboard/impl/data_quality/mock/test_providers/test_providers.tsx:
```diff
@@ -22,11 +24,24 @@ window.scrollTo = jest.fn();
     <I18nProvider>
       <ThemeProvider theme={() => ({ eui: euiDarkVars, darkMode: true })}>
-        <DataQualityProvider httpFetch={http.fetch}>{children}</DataQualityProvider>
+        <AssistantProvider
```
What will happen with Assistant functionality if we use data quality without that provider? Do we have a check somewhere?

spong replied:
We've got a test for this one: 

https://github.com/elastic/kibana/blob/542f3e5ed5b5df18b4425aa439336e420b4ff542/x-pack/packages/kbn-elastic-assistant/impl/assistant_context/index.test.tsx#L36-L42

jbudz reviewed (approved):
`packages/kbn-babel-preset` LGTM

elasticmachine commented:
Pinging @elastic/security-solution (Team: SecuritySolution)

YulNaumenko commented on x-pack/plugins/security_solution/public/common/components/event_details/alert_summary_view.tsx:
```diff
@@ -7,12 +7,10 @@
-
 import type { TimelineEventsDetailsItem } from '../../../../common/search_strategy';
-
+import type { BrowserFields } from '../../../../common/search_strategy/index_fields';
```
This changes is not related to the PR context.

YulNaumenko commented on x-pack/plugins/security_solution/public/common/components/event_details/summary_view.test.tsx:
```diff
@@ -53,6 +53,7 @@ const enrichedHostIpData: AlertSummaryRow['description'] = {
 };
 
 const mockCount = 90019001;
+
```
Could be excluded from PR as well

YulNaumenko commented on x-pack/plugins/security_solution/public/common/components/local_storage/index.tsx:
```diff
@@ -37,10 +38,18 @@ export const useLocalStorage = <T,>({
-
-      _setValue(value);
+    (value: T | ((prev: T) => T)) => {
+      if (typeof value === 'function') {
```
Where this is used and how this changes is related to the current PR?

spong replied:
It's used in the `useConversationStore` wrapper to store state in localstorage. `_setValue` needed to be extend to support providing the `prev` value as a function param when updating:

https://github.com/elastic/kibana/blob/542f3e5ed5b5df18b4425aa439336e420b4ff542/x-pack/plugins/security_solution/public/assistant/use_conversation_store/index.tsx#L19-L26

YulNaumenko reviewed (approved):
Tested locally and all is good. Have some minor comment related to the code and some to the user workflows. It could be done as a follow up work.
Amazing work!!!

kibanamachine commented:
## 💔 All backports failed

| Status | Branch | Result |
|:------:|:------:|:------|
|❌|8.8|Backport failed because of merge conflicts|

### Manual backport
To create the backport manually run:
```
node scripts/backport --pr 156933
```

### Questions ?
Please refer to the [Backport tool documentation](https://github.com/sqren/backport)

<!--- Backport version: 8.9.7 -->

spong commented:
## 💚 All backports created successfully

| Status | Branch | Result |
|:------:|:------:|:------|
|✅|8.8|[<img src="https://img.shields.io/github/pulls/detail/state/elastic/kibana/158974">](https://github.com/elastic/kibana/pull/158974)|

Note: Successful backport PRs will be merged automatically after passing CI.

### Questions ?
Please refer to the [Backport tool documentation](https://github.com/sqren/backport)

<!--- Backport version: 8.9.7 -->

nicpenning commented:
This is looking rad. Is this something that can be tested via feature flag as mentioned above in 8.8.1 or come later in 8.9.0?

Also, will this require OpenAI and/or Azure OpenAI subscriptions to operate? Looking to understand some pre-requisites so when this drops we can immediately test out.

This looks like a force multiplier for us analysts out there.

jamesspi commented:
@nicpenning , Answered you on the community slack :)
